PanelConfig: Add option to calculate min/max per field instead of using the global min/max in the data frame.

oscarkilhed opened the pull request:
**What is this feature?**

This let's the user decide if the auto min/min max for a field should be calculated per field or for all fields.

It's essentially does the same as https://github.com/grafana/grafana/pull/75803 but implements it as a configurable property on `FieldConfig`. This makes it possible to use it for all visualizations.


**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #26383

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

oscarkilhed committed:
Add option to calculate min max per field

oscarkilhed committed:
Fix eslint warnings

oscarkilhed commented on public/app/features/dashboard/components/PanelEditor/getVisualizationOptions.tsx:
```diff
@@ -97,16 +97,14 @@ export function getVisualizationOptions(props: OptionPaneRenderProps): OptionsPa
-
-    if (fieldOption.hideFromDefaults) {
-      continue;
+    if (fieldOption.isCustom) {
+      if (fieldOption.showIf && !fieldOption.showIf(currentFieldConfig.defaults.custom, data?.series)) {
+        continue;
+      }
+    } else {
+      if (fieldOption.showIf && !fieldOption.showIf(currentFieldConfig.defaults, data?.series)) {
+        continue;
+      }
```
This is a bit funky. `showIf` receiving the custom property if the option is party of the `custom` property, and the field config if it's not.

Also, it might be nicer to have this option disabled in the UI when both min and max are set, rather than having it hidden.

mckn replied:
Agree with you on this one. Maybe something we can address in a follow up PR?

dprokop replied:
Agree with you both. Would be great if we could improve `showIf` to also include standard properties. Def work to a separate PR and probably a breaking change to the API.

torkelo commented:
I think this looks ok, it's new option for almost every visualisation but can't find another way that adds this in a somewhat intuitive & consistent maner

oscarkilhed committed:
Add back hideFromDefaults that went missing

oscarkilhed committed:
whitespace

mckn commented on packages/grafana-data/src/field/fieldOverrides.ts:
```diff
@@ -168,12 +169,21 @@ export function applyFieldOverrides(options: ApplyFieldOverrideOptions): DataFra
       // Set the Min/Max value automatically
       let range: NumericRange | undefined = undefined;
       if (field.type === FieldType.number) {
-        if (!globalRange && (!isNumber(config.min) || !isNumber(config.max))) {
-          globalRange = findNumericFieldMinMax(options.data!);
+        if (!isNumber(config.min) || !isNumber(config.max)) {
+          if (config.localMinMax) {
+            const localRange = getMinMaxAndDelta(field);
+            const min = config.min ?? localRange!.min;
+            const max = config.max ?? localRange!.max;
+            range = { min, max, delta: max! - min! };
+          } else {
+            if (!globalRange) {
+              globalRange = findNumericFieldMinMax(options.data!);
+            }
+            const min = config.min ?? globalRange!.min;
+            const max = config.max ?? globalRange!.max;
+            range = { min, max, delta: max! - min! };
+          }
         }
```
Nit: it would have been nice to move this section to its own function so you can do early exits, reduce nesting and define the range as const.

mckn reviewed (approved):
LGTM! I have been looking for this functionality (as a user) a couple of times. Great that you are adding it.

Added some small nits to consider. Adding some tests would improve this PR even more.

oscarkilhed committed:
Add tests

oscarkilhed committed:
Refactor range calculation

codeincarnate reviewed (changes requested):
In general appears to be working. There does appear to be an issue when using the option in field overrides, and it will apply to multiple fields. See the video here:

https://github.com/grafana/grafana/assets/199847/de959c5d-3010-4a2a-81e9-d9226bb0ab16

oscarkilhed commented:
> In general appears to be working. There does appear to be an issue when using the option in field overrides, and it will apply to multiple fields. See the video here:
> 
>  Screen.Recording.2023-10-06.at.5.25.37.PM.mov

@codeincarnate I wasn't able to reproduce this, can you share some details on the setup?

![Peek 2023-10-09 10-02](https://github.com/grafana/grafana/assets/468940/f9136abb-2506-4178-a25d-a21b8e2ef1cb)

oscarkilhed committed:
Merge branch 'main' into oscark/local-min-max-fieldconfig

oscarkilhed commented:
@dprokop @ryantxu

Would like to hear your thoughts on this.

dprokop commented on packages/grafana-data/src/utils/tests/mockStandardProperties.ts:
```diff
@@ -67,6 +67,23 @@ export const mockStandardProperties = () => {
+  const localMinMax = {
+    id: 'localMinMax',
+    path: 'localMinMax',
+    name: 'localminmax',
```
I somehow don't like the `local...` prefix to this name. What about `fieldMinMax` instead? A little bit more descriptive. Thinking about current `min`/`max` it's not clear that it's dataset's min/max. Maybe let's avoid these problem when introducing the new property :)

oscarkilhed replied:
That is a good point. Will fix that.

dprokop reviewed (approved):
Overall I think this looks good. Can't think of any way we could introduce this without introducing a new standard option :(

codeincarnate reviewed (approved):
@oscarkilhed I've tried again and I've been unable to reproduce myself. Seems like it was just a heisenbug 😄. LGTM!

oscarkilhed committed:
Rename localMinMax -> fieldMinMax

oscarkilhed committed:
Remove the lint exceptions

Removing these as to not hide these once we get around to fixing the
underlying typing issue.

oscarkilhed committed:
Update docs
